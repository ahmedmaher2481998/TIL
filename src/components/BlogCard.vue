<script setup lang="ts">
import { TagChip } from '@/components'
import { type BlogType } from '@/types/index'
import { Card, CardDescription, CardFooter, CardHeader, CardTitle } from '@/components/ui/card'
import { formatDisplayDate, getInitials } from '@/utils'
import { Avatar, AvatarFallback, AvatarImage } from '@/components/ui/avatar'
import { CalendarTodayOutlined, ArrowUpwardOutlined } from '@vicons/material'
import { ref } from 'vue'
const { blog } = defineProps<{ blog: BlogType }>()
const ago = ref(formatDisplayDate(blog.createdat))
</script>

<template lang="html">
  <Card
    class="bg-green-100 text-white max-w-screen-lg relative overflow-hidden max-h-[500px] mx-2 md:mx-4"
  >
    <img class="bg-cover w-full rounded-lg" :src="blog.image" alt="{{ blog..title }}" />
    <div class="absolute bottom-0 glassy-feel w-full">
      <CardHeader>
        <!-- title with tldr  -->
        <div class="flex justify-between items-center w-full h-auto">
          <CardTitle class="font-semibold">{{ blog.title }}</CardTitle>
          <span
            class="h-10 text-sm w-10 rotate-45 hover:translate-x-1 hover:-translate-y-1 transition"
          >
            <ArrowUpwardOutlined class="font-thin" />
          </span>
        </div>
        <CardDescription class="text-slate-100 truncate">{{ blog.tldr }}</CardDescription>
      </CardHeader>

      <CardFooter class="flex items-center text-sm justify-between">
        <!-- author avatar with name  -->
        <div class="flex items-center justify-start">
          <div class="flex items-center justify-start gap-3">
            <Avatar>
              <AvatarImage :src="blog.user.avatar" :alt="blog.user.name" />
              <AvatarFallback>
                <!-- get initials  -->
                {{ getInitials(blog?.user.name ?? 'user name') }}
              </AvatarFallback>
            </Avatar>
            <span>{{ blog.user.name }}</span>
          </div>
          <div class="flex justify-center items-center gap-3 ms-4">
            <div class="h-10 p-2 border rounded-full w-10">
              <CalendarTodayOutlined />
            </div>
            <span> {{ ago }} </span>
          </div>
        </div>
        <!-- tags  -->
        <div class="flex gap-2">
          <TagChip
            v-for="tag in blog.tags"
            variant="glassy"
            :name="tag.title"
            :click="{
              href: '/tags/' + tag.id,
              linked: true
            }"
            :key="tag.id"
          />
        </div>
      </CardFooter>
    </div>
  </Card>
</template>

<style lang=""></style>
