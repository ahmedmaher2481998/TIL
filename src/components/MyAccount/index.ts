export { default as BlogsReadByUser } from './BlogsReadByUser.vue'
export { default as BlogsCreatedByUser } from './BlogsCreatedByUser.vue'
export { default as BlogCardSkeleton } from './BlogCard.Skeleton.vue'
export { default as UserProfile } from './UserProfile.vue'