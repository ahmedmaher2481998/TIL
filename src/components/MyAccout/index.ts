export { default as BlogsReadByUser } from './BlogsReadByUser.vue'
export { default as BlogsCreatedByUser } from './BlogsCreatedByUser.vue'
export { default as UserProfile } from './UserProfile.vue'