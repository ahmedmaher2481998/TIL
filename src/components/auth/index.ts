export { default as LoginPopUp } from './login/logInPopUp.vue'
export { default as LogoutPopUp } from './logout/logoutPopUp.vue'
export { default as RegisterPopUp } from './register/RegisterPopUp.vue'
export { default as UserMenu } from './userMenu/UserMenu.vue'
export { default as UserAvatarDisplay } from './userMenu/UserAvatarDisplay.vue'
export { default as LoginWithGoogleComponent } from './OAuth/logInWithGoogle.component.vue'
