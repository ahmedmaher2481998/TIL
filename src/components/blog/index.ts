export { default as BlogCard } from './BlogCard.vue'
export { default as TagChip } from './TagChip.vue'
export { default as FormInputField } from './createNewBlog/FormInputField.vue'
export { default as TagSelector } from './createNewBlog/TagSelector.vue'
export { default as CreateNewTagPopOver } from './createNewBlog/CreateNewTag.vue'
export { default as MarkDownContentInput } from './createNewBlog/MarkDownContentInput.vue'
export { default as TagChipComponent } from './DisplaySingleBlog/TagChip.component.vue'
