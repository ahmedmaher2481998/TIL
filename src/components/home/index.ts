export { default as FeaturedSectionComponent } from './FeaturedBlogSection/FeaturedSection.component.vue'
export { default as MainFeaturedBlogComponent } from './FeaturedBlogSection/MainFeaturedBlog.component.vue'
export { default as MainFeatureBlogSkeleton } from './FeaturedBlogSection/MainFeatureBlog.skeleton.vue'
export { default as SecondaryFeaturedBlogComponent } from './FeaturedBlogSection/SecondaryFeaturedBlog.component.vue'
export { default as SecondaryFeaturedBlogSkeleton } from './FeaturedBlogSection/SecondaryFeaturedBlog.skeleton.vue'
export { default as BlogsGrid } from './blogsGrid/BlogsGrid.vue'
export { default as BlogsGridSkeleton } from './blogsGrid/BlogsGrid.skeleton.vue'
