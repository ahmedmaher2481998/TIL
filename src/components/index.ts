// Lib components
export * from './ui/avatar'
export * from './ui/badge'
export { default as Button } from './ui/button/Button.vue'
export * from './ui/button'
export * from './ui/card'
export * from './ui/avatar'
export * from './ui/badge'
export * from './ui/dialog'
export * from './ui/input'
export * from './ui/label'
export * from './ui/command'
export * from './ui/form'
export * from './ui/popover'
export * from './ui/toast'
export * from './ui/skeleton'
export * from './ui/menubar'
export * from './ui/dropdown-menu'
// Main components [Shared ]
export { default as BlogCard } from './BlogCard.vue'
export { default as TagChip } from './TagChip.vue'
export { default as NavBar } from './NavBar.vue'
export { default as PopUp } from './PopUp.vue'
export { default as FormInputField } from './createNewBlog/FormInputField.vue'
export { default as ImageUpload } from './ImageUpload.vue'
// Create New Blog Components
export { default as TagSelector } from './createNewBlog/TagSelector.vue'
export { default as CreateNewTagPopOver } from './createNewBlog/CreateNewTag.vue'
export { default as MarkDownContentInput } from './createNewBlog/MarkDownContentInput.vue'

// auth components
export { default as LoginPopUp } from './auth/login/logInPopUp.vue'
export { default as LogoutPopUp } from './auth/logout/logoutPopUp.vue'
export { default as RegisterPopUp } from './auth/register/RegisterPopUp.vue'
export { default as UserMenu } from './auth/UserMenu.vue'
export { default as UserAvatarDisplay } from './auth/UserAvatarDisplay.vue'
