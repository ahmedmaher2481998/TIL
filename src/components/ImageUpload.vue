<script setup lang="ts">
import { Label, Input, FormItem } from '@/components'
import { UploadOutlined } from '@vicons/material'
import { ref, watch, toRefs, defineProps, withDefaults } from 'vue'
import { useField } from 'vee-validate'
import type { ClassValue } from 'clsx'
import clsx from 'clsx'

type propsType = {
  title: string
  name: string
  displayClasses?: ClassValue
}
const props = withDefaults(defineProps<propsType>(), {
  displayClasses: ''
})
const { title, name } = toRefs(props)

const { handleBlur, handleChange, errorMessage, value } = useField<File | null>(name, undefined, {
  initialValue: null
})
const blogCover = ref<HTMLImageElement | null>(null)
// display image when it's selected
watch(value, () => {
  if (value.value) {
    const reader = new FileReader()
    reader.onload = function (e: ProgressEvent<FileReader>) {
      blogCover!.value!.src = (e?.target?.result as string) ?? ''
    }
    reader.readAsDataURL(value.value)
  }
})
</script>
<template lang="html">
  <FormItem class="grid justify-start w-full items-center gap-1.5">
    <Label :for="name" class="mb-2 bg-" :class="{ 'text-red-600': Boolean(errorMessage) }">
      {{ title }}</Label
    >
    <label
      class="relative w-auto flex border p-2 rounded-lg cursor-pointer items-center"
      :for="name"
    >
      <span v-if="!value" class="flex items-center justify-center px-2">
        <UploadOutlined class="size-6 text-muted-foreground" />
      </span>
      <span class="flex flex-col">
        <img
          v-if="value"
          src=""
          ref="blogCover"
          class="shadow-lg rounded-lg object-cover"
          :class="displayClasses ? displayClasses : 'w-full h-auto'"
        />
        <span class="text-sm truncate text-gray-600 mt-4 text-center">
          {{ value ? value.name : 'Choose image' }}
        </span>
      </span>

      <Input
        accept="image/*"
        :id="name"
        type="file"
        class="pl-10 cursor-pointer hidden bg-green-600"
        @change="handleChange"
        @blur="handleBlur"
      />
    </label>
    <p v-if="Boolean(errorMessage)" class="text-sm font-medium text-destructive">
      {{ errorMessage }}
    </p>
  </FormItem>
</template>
