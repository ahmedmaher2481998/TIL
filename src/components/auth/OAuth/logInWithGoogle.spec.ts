import { flushPromises, mount, VueWrapper } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { Button, LoginWithGoogleComponent } from '@/components'
import { useAuth } from '@/stores'
import { beforeEach, describe, expect, it, vi } from 'vitest'

const loginWithGoogleMock = vi.fn(async () => {
  return Promise.resolve('logged in')
})
describe('MyComponent', () => {
  let wrapper: VueWrapper
  beforeEach(() => {
    wrapper = mount(LoginWithGoogleComponent)
    vi.mock('@/stores', () => ({
      useAuth: vi.fn(() => ({
        loginWithGoogle: loginWithGoogleMock,
      })),
    }))
    setActivePinia(createPinia())
  })

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  })


  it('renders the button with correct text and icon', () => {
    const button = wrapper.find('button')
    expect(button.exists()).toBe(true);
    expect(button.text()).toContain('Login with Google')
    // testing that the logo is rendered correctly
    const icon = wrapper.find<HTMLImageElement>('img[alt="google logo"]')
    expect(icon.exists()).toBe(true);
    expect(icon.attributes('src')).toBe('https://www.svgrepo.com/show/475656/google-color.svg')


  })

  it('calls auth.loginWithGoogle when the button is clicked', async () => {
    const button = wrapper.findComponent(Button)
    await button.trigger('click')
    await flushPromises()
    expect(loginWithGoogleMock).toHaveBeenCalled() //Fails 
  })

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
})