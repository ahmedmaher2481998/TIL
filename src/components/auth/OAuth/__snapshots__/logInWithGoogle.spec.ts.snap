// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyComponent > Renders Correctly 1`] = `"<button class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary hover:bg-primary/90 h-10 px-4 py-2 flex gap-2 rounded-lg text-muted w-full hover:border-slate-200 hover:shadow transition duration-150"><img class="w-6 h-6" src="https://www.svgrepo.com/show/475656/google-color.svg" loading="lazy" alt="google logo"><span>Login with Google</span></button>"`;

exports[`MyComponent > matches the snapshot 1`] = `"<button class="items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary hover:bg-primary/90 h-10 px-4 py-2 flex gap-2 rounded-lg text-muted w-full hover:border-slate-200 hover:shadow transition duration-150"><img class="w-6 h-6" src="https://www.svgrepo.com/show/475656/google-color.svg" loading="lazy" alt="google logo"><span>Login with Google</span></button>"`;
